File: frontend/js/payments.js
/**
 * Vexis Payment Module — Razorpay Integration
 * Handles subscription checks, pricing modal, and payment flow.
 */
import { getToken } from './api.js';

const API = 'https://vexis-backend-kklg.onrender.com/api';
const RAZORPAY_KEY = 'rzp_test_Sp78k3JIxzjZH1';

// ─── Plans config (mirrors backend) ────────────────────────────────────────
const PLANS = [
  {
    id:          'single',
    name:        'Single Report',
    price:       49,
    icon:        '📄',
    duration:    '1 Use',
    feature:     '1 AI Analysis or PDF Report',
    color:       '#34d399',
    badge:       null,
  },
  {
    id:          'explorer',
    name:        'Explorer',
    price:       99,
    icon:        '🔍',
    duration:    '7 Days',
    feature:     'Unlimited Analyses & PDF Reports',
    color:       '#60a5fa',
    badge:       null,
  },
  {
    id:          'pro',
    name:        'Pro',
    price:       199,
    icon:        '⚡',
    duration:    '30 Days',
    feature:     'Unlimited Analyses & PDF Reports',
    color:       '#a78bfa',
    badge:       null,
  },
  {
    id:          'elite',
    name:        'Elite',
    price:       499,
    icon:        '👑',
    duration:    '1 Year',
    feature:     'Unlimited Everything — Best Value',
    color:       '#f59e0b',
    badge:       'BEST VALUE',
  },
];

// ─── State ──────────────────────────────────────────────────────────────────
let _cachedStatus    = null;
let _cacheExpiry     = 0;
let _pendingCallback = null;     // Called after successful payment

// ─── Check subscription status (cached 60s) ─────────────────────────────────
export async function checkSubscription(force = false) {
  const now = Date.now();
  if (!force && _cachedStatus && now < _cacheExpiry) {
    return _cachedStatus;
  }
  try {
    const token = await getToken();
    const res   = await fetch(`${API}/payments/status`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    const data = await res.json();
    _cachedStatus = data;
    _cacheExpiry  = now + 60_000;   // Cache for 60 seconds
    return data;
  } catch (e) {
    console.error('[Payments] checkSubscription error:', e);
    return { active: false, plan: 'free' };
  }
}

// ─── Invalidate cache (call after payment) ──────────────────────────────────
export function invalidateSubCache() {
  _cachedStatus = null;
  _cacheExpiry  = 0;
}

// ─── Gate a feature: checks sub, shows modal if not active ──────────────────
/**
 * Call before any paid feature.
 * @param {Function} onAllowed - Called immediately if subscription is active,
 *                                or after successful payment.
 */
export async function requireSubscription(onAllowed) {
  const status = await checkSubscription();
  if (status.active) {
    onAllowed();
    return;
  }
  // Not subscribed — show pricing modal
  _pendingCallback = onAllowed;
  showPricingModal();
}

// ─── Build & show the pricing modal ─────────────────────────────────────────
export function showPricingModal() {
  // Remove any existing modal
  document.getElementById('vexis-payment-modal')?.remove();

  const overlay = document.createElement('div');
  overlay.id        = 'vexis-payment-modal';
  overlay.className = 'payment-overlay';
  overlay.innerHTML = `
    <div class="payment-modal" role="dialog" aria-modal="true" aria-label="Choose a plan">

      <div class="payment-modal-header">
        <button class="payment-modal-close" id="pay-modal-close" aria-label="Close">✕</button>
        <div class="payment-modal-icon">🔐</div>
        <h2 class="payment-modal-title">Unlock AI Intelligence</h2>
        <p class="payment-modal-subtitle">Choose a plan to access AI Health Analysis & PDF Reports</p>
      </div>

      <!-- Plans grid -->
      <div class="plans-grid" id="plans-grid">
        ${PLANS.map(p => `
          <div class="plan-card" data-plan="${p.id}" id="plan-card-${p.id}" tabindex="0" role="button" aria-label="Select ${p.name} plan">
            ${p.badge ? `<div class="plan-badge">${p.badge}</div>` : ''}
            <div class="plan-icon">${p.icon}</div>
            <div class="plan-name">${p.name}</div>
            <div class="plan-price">
              <span class="currency">₹</span>
              <span class="amount">${p.price}</span>
            </div>
            <div class="plan-duration">${p.duration}</div>
            <div class="plan-feature">${p.feature}</div>
            <button class="plan-cta" data-plan="${p.id}">Get ${p.name}</button>
          </div>
        `).join('')}
      </div>

      <!-- Processing state -->
      <div class="payment-processing" id="pay-processing">
        <div class="payment-spinner"></div>
        <p style="color:rgba(255,255,255,0.6); margin:0;">Processing payment securely…</p>
      </div>

      <!-- Success state -->
      <div class="payment-success" id="pay-success">
        <div class="payment-success-icon">✅</div>
        <h3>Payment Successful!</h3>
        <p>Your subscription is now active. Loading your analysis…</p>
      </div>

      <div class="payment-footer">
        <span>🔒 Secured by</span>
        <span style="font-weight:600;color:rgba(255,255,255,0.5)">Razorpay</span>
        <span>· Test Mode Active</span>
      </div>

    </div>
  `;

  document.body.appendChild(overlay);

  // Animate in
  requestAnimationFrame(() => overlay.classList.add('active'));

  // Close on overlay click or close button
  document.getElementById('pay-modal-close').addEventListener('click', closeModal);
  overlay.addEventListener('click', e => { if (e.target === overlay) closeModal(); });
  // Close on Escape
  const escHandler = e => { if (e.key === 'Escape') closeModal(); };
  document.addEventListener('keydown', escHandler);
  overlay._escHandler = escHandler;

  // Plan CTA clicks
  overlay.querySelectorAll('.plan-cta').forEach(btn => {
    btn.addEventListener('click', e => {
      e.stopPropagation();
      initiatePayment(btn.dataset.plan);
    });
  });
  // Also allow clicking the card itself
  overlay.querySelectorAll('.plan-card').forEach(card => {
    card.addEventListener('keydown', e => {
      if (e.key === 'Enter' || e.key === ' ') initiatePayment(card.dataset.plan);
    });
  });
}

function closeModal() {
  const overlay = document.getElementById('vexis-payment-modal');
  if (!overlay) return;
  if (overlay._escHandler) document.removeEventListener('keydown', overlay._escHandler);
  overlay.classList.remove('active');
  setTimeout(() => overlay.remove(), 350);
  _pendingCallback = null;
}

function showProcessing() {
  document.getElementById('plans-grid').style.display      = 'none';
  document.getElementById('pay-processing').classList.add('active');
  document.getElementById('pay-success').classList.remove('active');
}

function showSuccess() {
  document.getElementById('pay-processing').classList.remove('active');
  document.getElementById('pay-success').classList.add('active');
}

// ─── Initiate Razorpay checkout ──────────────────────────────────────────────
async function initiatePayment(planId) {
  showProcessing();
  try {
    const token = await getToken();

    // 1. Create order on backend
    const orderRes = await fetch(`${API}/payments/create-order`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body:    JSON.stringify({ plan_id: planId }),
    });
    const order = await orderRes.json();

    if (!orderRes.ok || !order.order_id) {
      throw new Error(order.error || 'Failed to create order');
    }

    // 2. Hide processing, open Razorpay checkout
    document.getElementById('pay-processing').classList.remove('active');
    document.getElementById('plans-grid').style.display = 'grid';

    const options = {
      key:         order.key_id,        // ← use the SAME key the backend used to create the order
      amount:      order.amount,
      currency:    order.currency,
      name:        'Vexis AI',
      description: order.plan_name,
      image:       '',
      order_id:    order.order_id,
      prefill: {},
      theme:       { color: '#818cf8' },
      modal: {
        ondismiss: () => {
          // User closed checkout — reset modal back to plans
          document.getElementById('plans-grid').style.display = 'grid';
          document.getElementById('pay-processing').classList.remove('active');
        },
      },
      handler: async (response) => {
        await handlePaymentSuccess(response, planId);
      },
    };

    // Load Razorpay SDK lazily if not already loaded
    await loadRazorpaySdk();
    const rzp = new window.Razorpay(options);
    rzp.open();

  } catch (err) {
    console.error('[Payments] initiatePayment error:', err);
    document.getElementById('plans-grid').style.display = 'grid';
    document.getElementById('pay-processing').classList.remove('active');
    showToast(`Payment error: ${err.message}`, 'error');
  }
}

// ─── Handle successful payment ───────────────────────────────────────────────
async function handlePaymentSuccess(response, planId) {
  showProcessing();
  try {
    const token = await getToken();
    const res   = await fetch(`${API}/payments/verify`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body:    JSON.stringify({
        razorpay_payment_id: response.razorpay_payment_id,
        razorpay_order_id:   response.razorpay_order_id,
        razorpay_signature:  response.razorpay_signature,
        plan_id:             planId,
      }),
    });
    const result = await res.json();

    if (!res.ok || !result.success) {
      throw new Error(result.error || 'Verification failed');
    }

    // Invalidate cache so next check fetches fresh status
    invalidateSubCache();

    showSuccess();

    // After 1.8s — close modal and trigger the pending action
    setTimeout(() => {
      closeModal();
      if (typeof _pendingCallback === 'function') {
        const cb = _pendingCallback;
        _pendingCallback = null;
        cb();
      }
    }, 1800);

  } catch (err) {
    console.error('[Payments] handlePaymentSuccess error:', err);
    showToast(`Verification error: ${err.message}`, 'error');
    document.getElementById('pay-processing').classList.remove('active');
    document.getElementById('plans-grid').style.display = 'grid';
  }
}

// ─── Lazily load Razorpay SDK ────────────────────────────────────────────────
function loadRazorpaySdk() {
  return new Promise((resolve, reject) => {
    if (window.Razorpay) { resolve(); return; }
    const s  = document.createElement('script');
    s.src    = 'https://checkout.razorpay.com/v1/checkout.js';
    s.onload = resolve;
    s.onerror = () => reject(new Error('Failed to load Razorpay SDK'));
    document.head.appendChild(s);
  });
}

// ─── Small toast helper (uses existing toast if available) ───────────────────
function showToast(msg, type = 'info') {
  if (typeof window.showToast === 'function') {
    window.showToast(msg, type);
    return;
  }
  const t = document.createElement('div');
  t.textContent = msg;
  Object.assign(t.style, {
    position: 'fixed', bottom: '1.5rem', right: '1.5rem',
    background: type === 'error' ? '#ef4444' : '#818cf8',
    color: '#fff', padding: '0.75rem 1.25rem', borderRadius: '10px',
    fontFamily: 'Rajdhani, sans-serif', fontSize: '0.9rem',
    zIndex: '99999', boxShadow: '0 8px 30px rgba(0,0,0,0.4)',
    transition: 'opacity 0.3s', pointerEvents: 'none',
  });
  document.body.appendChild(t);
  setTimeout(() => { t.style.opacity = '0'; setTimeout(() => t.remove(), 300); }, 3000);
}

File: backend/routes/payments.py
"""
Vexis Payments Blueprint — Razorpay Integration
Handles subscription plan creation, payment verification, and status checks.
"""
import os
import hmac
import hashlib
from datetime import datetime, timedelta
from flask import Blueprint, request, jsonify
from utils.firebase_auth import firebase_required

payments_bp = Blueprint('payments', __name__)

# ── Plan definitions ─────────────────────────────────────────────────────────
PLANS = {
    'single': {
        'name':         'Single Report',
        'amount':       4900,           # ₹49 in paise
        'duration_days': None,          # No expiry — just 1 use
        'reports':      1,
        'description':  '1 AI Analysis or PDF Report',
    },
    'explorer': {
        'name':         'Explorer',
        'amount':       9900,           # ₹99 in paise
        'duration_days': 7,
        'reports':      None,           # Unlimited
        'description':  'Unlimited for 7 Days',
    },
    'pro': {
        'name':         'Pro',
        'amount':       19900,          # ₹199 in paise
        'duration_days': 30,
        'reports':      None,
        'description':  'Unlimited for 30 Days',
    },
    'elite': {
        'name':         'Elite',
        'amount':       49900,          # ₹499 in paise
        'duration_days': 365,
        'reports':      None,
        'description':  'Unlimited for 1 Year',
    },
}


def _get_razorpay_client():
    import razorpay
    # Fallback to the test key if env var is missing
    key_id     = os.getenv('RAZORPAY_KEY_ID', 'rzp_test_Sp78k3JIxzjZH1')
    key_secret = os.getenv('RAZORPAY_KEY_SECRET', '')
    return razorpay.Client(auth=(key_id, key_secret))


def _fs():
    from firebase_admin import firestore
    return firestore.client()


def check_subscription(uid: str) -> dict:
    """
    Check if a user has an active subscription.
    Returns {'active': bool, 'plan': str, 'reports_remaining': int|None, 'valid_until': str|None}
    """
    try:
        doc = _fs().collection('users').document(uid) \
                   .collection('subscription').document('current').get()

        if not doc.exists:
            return {'active': False, 'plan': 'free'}

        data = doc.to_dict()
        plan = data.get('plan', 'free')

        if plan == 'single':
            remaining = data.get('reports_remaining', 0)
            if remaining and remaining > 0:
                return {'active': True, 'plan': plan, 'reports_remaining': remaining}
            return {'active': False, 'plan': 'free', 'reports_remaining': 0}

        # Time-based plans
        valid_until = data.get('valid_until')
        if valid_until:
            try:
                expiry = datetime.fromisoformat(valid_until)
                if datetime.utcnow() < expiry:
                    return {
                        'active':      True,
                        'plan':        plan,
                        'valid_until': valid_until,
                        'reports_remaining': None,
                    }
            except Exception:
                pass

        return {'active': False, 'plan': 'free'}

    except Exception as e:
        print(f'[PAYMENTS] check_subscription error: {e}')
        return {'active': False, 'plan': 'free'}


def consume_single_report(uid: str):
    """Decrement reports_remaining for single-plan users. Call after successful analysis."""
    try:
        ref = _fs().collection('users').document(uid) \
                   .collection('subscription').document('current')
        doc = ref.get()
        if doc.exists:
            data = doc.to_dict()
            if data.get('plan') == 'single':
                remaining = data.get('reports_remaining', 0)
                ref.update({'reports_remaining': max(0, remaining - 1)})
    except Exception as e:
        print(f'[PAYMENTS] consume_single_report error: {e}')


# ── POST /api/payments/create-order ─────────────────────────────────────────
@payments_bp.route('/payments/create-order', methods=['POST'])
@firebase_required
def create_order():
    """Create a Razorpay order and return the order_id + key_id to frontend."""
    try:
        body    = request.get_json() or {}
        plan_id = body.get('plan_id', '')

        if plan_id not in PLANS:
            return jsonify({'error': f'Invalid plan: {plan_id}'}), 400

        plan   = PLANS[plan_id]
        client = _get_razorpay_client()

        order = client.order.create({
            'amount':          plan['amount'],
            'currency':        'INR',
            'payment_capture': 1,
            'notes': {
                'plan_id': plan_id,
                'uid':     request.user['uid'],
            },
        })

        return jsonify({
            'order_id':  order['id'],
            'amount':    plan['amount'],
            'currency':  'INR',
            'plan_name': plan['name'],
            'key_id':    os.getenv('RAZORPAY_KEY_ID', 'rzp_test_Sp78k3JIxzjZH1'),
        }), 200

    except Exception as e:
        print(f'[PAYMENTS] create_order error: {e}')
        return jsonify({'error': str(e)}), 500


# ── POST /api/payments/verify ────────────────────────────────────────────────
@payments_bp.route('/payments/verify', methods=['POST'])
@firebase_required
def verify_payment():
    """
    Verify Razorpay signature and activate the subscription in Firestore.
    Called by frontend after Razorpay checkout success callback.
    """
    try:
        body         = request.get_json() or {}
        payment_id   = body.get('razorpay_payment_id', '')
        order_id     = body.get('razorpay_order_id', '')
        signature    = body.get('razorpay_signature', '')
        plan_id      = body.get('plan_id', '')

        if not all([payment_id, order_id, signature, plan_id]):
            return jsonify({'error': 'Missing required payment fields'}), 400

        if plan_id not in PLANS:
            return jsonify({'error': 'Invalid plan'}), 400

        # ── Verify HMAC-SHA256 signature ─────────────────────────────────────
        secret = os.getenv('RAZORPAY_KEY_SECRET', '').encode()
        message = f'{order_id}|{payment_id}'.encode()
        expected_sig = hmac.new(secret, message, hashlib.sha256).hexdigest()

        if expected_sig != signature:
            return jsonify({'error': 'Payment signature verification failed'}), 400

        # ── Activate subscription in Firestore ───────────────────────────────
        plan = PLANS[plan_id]
        now  = datetime.utcnow()
        uid  = request.user['uid']

        sub_data = {
            'plan':                plan_id,
            'razorpay_payment_id': payment_id,
            'razorpay_order_id':   order_id,
            'activated_at':        now.isoformat(),
        }

        if plan['duration_days']:
            sub_data['valid_until']        = (now + timedelta(days=plan['duration_days'])).isoformat()
            sub_data['reports_remaining']  = None
        else:
            # Single-report plan
            sub_data['valid_until']        = None
            sub_data['reports_remaining']  = plan['reports']

        _fs().collection('users').document(uid) \
             .collection('subscription').document('current').set(sub_data)

        print(f'[PAYMENTS] Activated plan={plan_id} for uid={uid} payment={payment_id}')

        return jsonify({
            'success':      True,
            'plan':         plan_id,
            'plan_name':    plan['name'],
            'valid_until':  sub_data.get('valid_until'),
        }), 200

    except Exception as e:
        print(f'[PAYMENTS] verify_payment error: {e}')
        return jsonify({'error': str(e)}), 500


# ── GET /api/payments/status ─────────────────────────────────────────────────
@payments_bp.route('/payments/status', methods=['GET'])
@firebase_required
def payment_status():
    """Return current subscription status for the authenticated user."""
    uid    = request.user['uid']
    status = check_subscription(uid)
    return jsonify(status), 200


# ── GET /api/payments/plans ──────────────────────────────────────────────────
@payments_bp.route('/payments/plans', methods=['GET'])
def get_plans():
    """Return available plans (public endpoint, no auth needed)."""
    return jsonify({
        pid: {
            'name':        p['name'],
            'amount':      p['amount'],
            'amount_inr':  p['amount'] // 100,
            'description': p['description'],
        }
        for pid, p in PLANS.items()
    }), 200
